File: Python/house-robber-ii.py
class Solution:
    # @param {integer[]} nums
    # @return {integer}
    def rob1(self, num):
        n = len(num)
        even = odd = 0
        for i in xrange(n):
            #dp[i] = max(dp[i-1], dp[i-2]+num[i])
            if i%2==0:
                even = max(odd, even+num[i])
            else:
                odd = max(even, odd+num[i])

        return max(even, odd)
        
    def rob(self, nums):
        n = len(nums)
        if n==0: return 0
        if n==1: return nums[0]
        return max(self.rob1(nums[1:]), self.rob1(nums[:n-1]))

File: C++/house-robber-ii.cpp
class Solution {
private:
    int rob(std::vector<int> &nums, int start, int end) {
        int dp[2] = {0, 0};
        int flag = 0;
        for(int i=start; i<end; ++i) {
            dp[flag] = std::max(dp[1-flag], dp[flag]+nums[i]);
            flag = 1-flag;
        }
        return std::max(dp[0], dp[1]);
    }
public:
    int rob(std::vector<int> &nums) {
        int n = nums.size();
        if(n==0) return 0;
        if(n==1) return nums[0];
        return std::max(rob(nums, 0, n-1), rob(nums, 1, n));
    }
};
